analyzer.py

```diff
@@ -0,0 +1,320 @@
+        trial1_cam2 = cv2.VideoCapture(os.path.join(self._trials_path, trial1, trial1 + '_cam2.avi'))
+        trial2_cam2 = cv2.VideoCapture(os.path.join(self._trials_path, trial2, trial2 + '_cam2.avi'))
+        # Compare hashes
+        if self._config['cam1s_are_the_same_view']:
+            cam1_dif, noc1 = self._compare_two_videos(trial1_cam1, trial2_cam1)
+            cam2_dif, noc2 = self._compare_two_videos(trial1_cam2, trial2_cam2)
+        else:
+            cam1_dif, noc1 = self._compare_two_videos(trial1_cam1, trial2_cam2)
+            cam2_dif, noc2 = self._compare_two_videos(trial1_cam2, trial2_cam1)
+        return (cam1_dif + cam2_dif) / (noc1 + noc2)
+
+    def analyze_marker_similarity_project(self):
+        '''Analyze all videos in a project and get their average rhythmicity. This assumes that all cam1/2 pairs are either the same or different!'''
+        marker_similarity = {}
+        list_of_trials = [folder for folder in os.listdir(self._trials_path) if os.path.isdir(os.path.join(self._trials_path, folder)) and not folder.startswith('.')]
+        yaml = YAML()
+
+        trial_perms = combinations(list_of_trials, 2)
+        for trial1, trial2 in trial_perms:
+            self._config['trial_1_name'] = trial1
+            self._config['trial_2_name'] = trial2
+            with open(os.path.join(self._config['working_dir'], 'project_config.yaml'), 'w') as file:
+                yaml.dump(self._config, file)
+            # TODO: pass the trial names directly instead of writing into config
+            marker_similarity[(trial1, trial2)] = abs(self.analyze_marker_similarity_trial())
+        return marker_similarity
+
+    def analyze_marker_similarity_trial(self):
+        '''Analyze marker similarity for a pair of trials. Returns the mean difference for paired marker positions (X - X, Y - Y for each marker)'''
+        # Find CSVs for each trial
+        trial1 = self._config['trial_1_name']
+        trial2 = self._config['trial_2_name']
+        trial1_path = os.path.join(self._trials_path, trial1)
+        trial2_path = os.path.join(self._trials_path, trial2)
+        # Get a list of markers that each trial have in commmon
+        # Marker similarity is always in rgb mode.
+        bodyparts1 = self._data_processor.get_bodyparts_from_xma(trial1_path, mode='rgb')
+        bodyparts2 = self._data_processor.get_bodyparts_from_xma(trial2_path, mode='rgb')
+
+        markers_in_common = [marker for marker in bodyparts1 if marker in bodyparts2]
+        bodyparts_xy = [f'{marker}_X' for marker in markers_in_common] + [f'{marker}_Y' for marker in markers_in_common]
+        trial1_csv = pd.read_csv(os.path.join(trial1_path, trial1 + '.csv'))
+        trial2_csv = pd.read_csv(os.path.join(trial2_path, trial2 + '.csv'))
+
+        marker_similarity = sum((trial1_csv[marker] - trial2_csv[marker]).sum() / (len(trial1_csv[marker]) + len(trial2_csv[marker])) for marker in bodyparts_xy) / len(bodyparts_xy)
```

Comment:
It actually makes more sense to use the distance formula on this data... i.e. get the average XY position of a marker in one trial (trial1_csv[f'{marker}_X].mean(), trial1_csv[f'{marker}_Y].mean()) and then use the distance formula to get the average distance between two markers.

```
for marker in markers_in_common:
  avg_xy_t1[marker] = (trial1_csv[f'{marker}_X].mean(), trial1_csv[f'{marker}_Y].mean())
  # same for t2
  
  avg_distance[marker] = sqrt((avg_xy_t2[marker][0] - avg_xy_t1[marker][0])**2 + (avg_xy_t2[marker][1] - avg_xy_t1[marker][1])**2)
```

Reply:
Then you would get the grand mean of avg_distance if you really wanted to boil it down into one big number (or you could break down the similarity scores per marker.

Reply:
done (although pls feel free to correct if i got the math wrong)